### ISSUE TITLE
Improve training data

### ISSUE THREAD

**Issue by ariasanovsky** [ariasanovsky/azdopt#37]
We've spent a lot of time adjusting the internals, but need to check the quality of the data passed into the network. Cross entropy seems stuck close to $5.8$. With some `dbg!` and `prinln!` inspection, we seem to be passing in incorrect data into the network.

Primarily, we need to check

```rust
pub fn update_existing_nodes<N, Space>(
    self,
    c_t: &impl Cost<f32>,
    s_t: &Space::State,
    // s_t: &impl State,
    p_t: &'a P,
    probs_t: &[f32],
    g_star_theta_s_t: &[f32],
) -> Option<NewTreeLevel<P>>
where
    Space: StateActionSpace,
    P: ActionPathFor<Space> + Ord,
    N: TreeNode<Path = P, State = Space::State>,
    P: Ord + Clone,
{ /* ... */ }
```

To do this:

- [ ] Define a simple state-action space and implement `StateActionSpace` for it
- [ ] write tests for `simulate_once`
- [ ] write tests for `update_existing_nodes` for a specific transition vector

**Comment by ariasanovsky** [comment]
# Major update

The learning loop was only calling the optimizer on the `core_model` and thus only its weights and biases were updated. It helped to rewrite with

```rust
let mut core_model = dev.build_module::<Core, f32>();
let mut logits_model = dev.build_module::<Logits, f32>();
let mut value_model = dev.build_module::<Valuation, f32>();

let config = AdamConfig {
    lr: 1e-3,
    betas: [0.9, 0.999],
    eps: 1e-8,
    weight_decay: None, //Some(WeightDecay::L2(1e-6)), // Some(WeightDecay::Decoupled(1e-6)),
};
let mut core_optimizer = Adam::new(&core_model, config.clone());
let mut logits_optimizer = Adam::new(&logits_model, config.clone());
let mut value_optimizer = Adam::new(&value_model, config);
```

and manually check weights and biases. Here, we noticed that calling `core_optimizer.update` only updated the `core` weights and biases. I also have a patchy fix that performs

```rust
let root_tensor = dev.tensor(v_t);
let traced_predictions = core_model.forward(root_tensor.leaky_traced());
let predicted_logits = logits_model.forward(traced_predictions);
let cross_entropy =
    cross_entropy_with_logits_loss(predicted_logits, observed_probabilities_tensor.clone());
let entropy = cross_entropy.array();
let mut gradients = cross_entropy.backward();
core_optimizer.update(&mut core_model, &gradients)
    .expect("optimizer failed");
core_model.zero_grads(&mut gradients);
```

4 times, instead of only once or twice. I'm debating the following fixes.

## Maintain the Y-shaped model

Currently, our model evaluates as 
$$\begin{array}{l}
    \text{Space} & \Rightarrow & \text{Hidden Layer 1} & \Rightarrow & \text{Hidden Layer 2} & \Rightarrow & \text{Logits} \\
    & & & & & \Rightarrow & \text{Mean Max Gain} 
\end{array}$$

Since loss combines both $L_2$ and cross entropy, we need to update coefficients in $2$ places, corresponding to $2$ loss functions. Currently, we patched this by crudely calling optimizers $4$ times.

## Break the loss functions into separate networks
In other words, define two separate neural networks, for example
$$\begin{array}{l}
    \text{Space} & \Rightarrow & \text{Hidden Layer 1} & \Rightarrow & \text{Hidden Layer 2} & \Rightarrow & \text{Logits} \\
    \text{Space} & \Rightarrow & \text{Hidden Layer 1} & \Rightarrow & \text{Hidden Layer 2} & \Rightarrow & \text{Mean Max Gain} 
\end{array}$$
The disadvantage is we relearn the same patterns in the two separate hidden layers.

## Additional feature: penalize high costs
We may also wish to penalize models which do not improve the cost function. This can take a few forms:
- add a loss term which punishes unfavorable values for $g^\ast$, $\overline{g}^\ast$, or $c^\ast$.
    - $g^\ast$ or $c^\ast$ would require adding a dimension to the $\text{Mean Max Gain}$ space.
    - we only want to update the weights in $\text{Hidden Layer 2}\Rightarrow\text{Logits}$ for this.
